### PR TITLE
Support mkldnn for macos

### PR DESCRIPTION
**Pull request by lisa0314**
Fix [intel/webml-polyfill#490](https://github.com/intel/webml-polyfill/issues/490).
For Macos:
- run --use-mkldnn --no-sandbox for choosing MKL-DNN backend.
- run mps or bnns with no option

### CODE REVIEW

**Comment by lisa0314** [comment]
@fujunwei , please help to review.

File: services/ml/ml_service.cc [hunk 1]
```diff
@@ -28,13 +31,22 @@ MLService::~MLService() = default;
...
-
-#if defined(OS_LINUX) || defined(OS_WIN)
+#if defined(OS_MACOSX)
+  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  std::string value = "bnns";
+  if (command_line->HasSwitch(switches::kCPUType)) {
+    value = command_line->GetSwitchValueASCII(ml::switches::kCPUType);
+  }
+  if (value == "bnns") {
+    registry_.AddInterface(base::Bind(&NeuralNetworkImplMac::Create));
```
Comment (fujunwei) [inline services/ml/ml_service.cc]: MPS also need the interface.

File: services/ml/compilation_impl.cc [hunk 2]
```diff
@@ -49,7 +49,9 @@ void CompilationImpl::Finish(int32_t preference, FinishCallback callback) {
   DLOG(INFO) << "  "
              << "preference: " << preference;
   if (preference == mojom::PREFER_SUSTAINED_SPEED) {
-    delegate_ = std::make_unique<CompilationDelegateClDnn>(this);
+    #if defined(OS_WIN) || defined(OS_LINUX)
```
Comment (fujunwei) [inline services/ml/compilation_impl.cc]: Use `git cl format` to remove 4-space.

File: services/ml/BUILD.gn [hunk 3]
```diff
@@ -195,6 +221,24 @@ if (is_win || is_linux) {
       "$root_out_dir/{{source_file_part}}",
     ]
   }
+} else if (is_mac) {
+  mkldnn_dll_path = "//third_party/mkl-dnn/lib/${os_folder}/Release"
```
Comment (fujunwei) [inline services/ml/BUILD.gn]: Can we merge those copy function into a common function?

File: chrome/browser/about_flags.cc [hunk 4]
```diff
@@ -1099,6 +1100,14 @@ const FeatureEntry::FeatureVariation kWebXrRenderPathVariations[] = {
      arraysize(kWebXrRenderPathChoiceSharedBuffer), nullptr}};
 #endif  // defined(OS_ANDROID) && BUILDFLAG(ENABLE_VR)
 
+#if defined(OS_MACOSX)
+const FeatureEntry::Choice kServiceMLCPUBackendChoices[] = {
+    {"default", "", ""},
+    {"choose for bnns", ml::switches::kCPUType, "bnns"},
```
Comment (fujunwei) [inline chrome/browser/about_flags.cc]: It's better to define const char array in ml_switch.h

File: chrome/browser/about_flags.cc [hunk 5]
```diff
@@ -3204,6 +3214,12 @@ const FeatureEntry kFeatureEntries[] = {
      FEATURE_VALUE_TYPE(omnibox::kEnableClipboardProvider)},
 #endif
 
+#if defined(OS_MACOSX)
+    {"cpu-backend-for-ml", "choose one cpu backend for macos",
```
Comment (fujunwei) [inline chrome/browser/about_flags.cc]: Ditto.

File: services/ml/BUILD.gn [hunk 6]
```diff
@@ -86,6 +90,10 @@ source_set("lib") {
     libs = [ "neuralnetworks" ]
   } else if (is_mac) {
     sources += [
+      "compilation_delegate_mkl_dnn.cc",
```
Comment (fujunwei) [inline services/ml/BUILD.gn]: It's better to  merge the repetitive usage file with Linux section.

**Comment by lisa0314** [comment]
@fujunwei Update the codes according to your comments. Please help to review again. Thanks!

File: services/ml/ml_service.cc [hunk 7]
```diff
@@ -28,13 +31,24 @@ MLService::~MLService() = default;
 void MLService::OnStart() {
   ref_factory_.reset(new service_manager::ServiceContextRefFactory(
       context()->CreateQuitClosure()));
-
-#if defined(OS_LINUX) || defined(OS_WIN)
+#if defined(OS_MACOSX)
+  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  std::string value = "bnns";
+  if (command_line->HasSwitch(switches::kCPUType)) {
```
Comment (fujunwei) [inline services/ml/ml_service.cc]: std::string value = command_line->HasSwitch(switches::kCPUType) ?  command_line->GetSwitchValueASCII(ml::switches::kCPUType) : "mps";

**Comment by huningxin** [comment]
@lisa0314 , I have concerns about the binaries:
```
BIN +2.59 MB third_party/mkl-dnn/lib/Mac64/Release/libiomp5.dylib
BIN +12.9 MB third_party/mkl-dnn/lib/Mac64/Release/libmkldnn.0.17.0.dylib
BIN +12.9 MB third_party/mkl-dnn/lib/Mac64/Release/libmkldnn.0.dylib
BIN +12.9 MB third_party/mkl-dnn/lib/Mac64/Release/libmkldnn.dylib
BIN +88.6 MB third_party/mkl-dnn/lib/Mac64/Release/libmklml.dylib
```

Please avoid checking in multiple libmkldnn dynamic libs. As mklml is very big, please try to build mkldnn with iomp and icc (Intel compiler). And please compare the performance (image classification) for the two builds:
1. with iomp + mklml
2. with iomp + icc

Then we can decide to use mklml or not.

For you reference, for Linux build, I use iomp + icc whose performance is close to iomp + mklml.

**Comment by lisa0314** [comment]
@huningxin Thanks for your review. I will remove redundant libmkldnn dynamic libs and compare the performance for those two builds.

**Comment by lisa0314** [comment]
@huningxin , I have compared the performance between those two builds,  iomp + icc whose performance is also close to iomp + mklml as Linux build. So I suggest we don't need to use mklml and I have pushed new libmkldnn.dylib into the repo.

**Comment by huningxin** [comment]
@lisa0314 , thanks for the investigation and reduce the binary size!

I tried to build your branch and it fails to run with error message:
```
$ ./Chromium.app/Contents/MacOS/Chromium --cpu-backend-for-ml=mkldnn --no-sandbox
dlopen /Users/nhu/code/chromium-src/src/out/Release/./Chromium.app/Contents/MacOS/../Versions/70.0.3503.0/Chromium Framework.framework/Chromium Framework: dlopen(/Users/nhu/code/chromium-src/src/out/Release/./Chromium.app/Contents/MacOS/../Versions/70.0.3503.0/Chromium Framework.framework/Chromium Framework, 261): Library not loaded: @rpath/libmkldnn.0.dylib
  Referenced from: /Users/nhu/code/chromium-src/src/out/Release/Chromium.app/Contents/Versions/70.0.3503.0/Chromium Framework.framework/Versions/A/Chromium Framework
  Reason: image not found
Abort trap: 6
```

Did I miss anything?

**Comment by lisa0314** [comment]
@huningxin , this issue didn't show up for me. And I will try to find the reasons as soon as possible.

**Comment by lisa0314** [comment]
@huningxin ,I have updated the PR in two points:
- Update the name of dylib from libmkldnn.dylib to libmkldnn.0.dylib
- Add the rpath value for runtime linking if it's non_component_build.
Please help to try again. Thanks!

**Comment by huningxin** [comment]
I tried 43f5eb7658fc8ecfa8a728433763e1e4edf2c4e5 and release build. There is still an error:
```
./Chromium.app/Contents/MacOS/Chromium --cpu-backend-for-ml=bnns --no-sandbox
dlopen /Users/nhu/code/chromium-src/src/out/Release/./Chromium.app/Contents/MacOS/../Versions/70.0.3503.0/Chromium Framework.framework/Chromium Framework: dlopen(/Users/nhu/code/chromium-src/src/out/Release/./Chromium.app/Contents/MacOS/../Versions/70.0.3503.0/Chromium Framework.framework/Chromium Framework, 261): Library not loaded: @rpath/libmkl_rt.dylib
  Referenced from: /Users/nhu/code/chromium-src/src/out/Release/libmkldnn.0.dylib
  Reason: image not found
Abort trap: 6

```

**Comment by fujunwei** [comment]
I think the root cause of it is the library isn't packaged in the binary.

Chromium.app\Contents\Versions\70.0.3503.0\Chromium Framework.framework\Libraries the directory hasn't mkldnn library.

File: chrome/BUILD.gn [hunk 8]
```diff
@@ -1281,6 +1302,11 @@ if (is_win) {
         ":chrome_dll",
       ]
     }
+
+    ldflags += [
```
Comment (fujunwei) [inline chrome/BUILD.gn]: Does this section still need?

Comment (lisa0314) [reply chrome/BUILD.gn]: Yes, it's for non-component build.

**Comment by huningxin** [comment]
I managed to build and try. There are a few issues:
1. it reports `Library not loaded: @rpath/libiomp5.dylib` when run with command line `./Chromium.app/Contents/MacOS/Chromium --cpu-backend-for-ml=mkldnn --no-sandbox` (I can workaround it by inserting `LD_LIBRARY_PATH=`, however it needs either to be documented or use a wrapper script, e.g. chrome-wrapper for Linux)
2. When running with switch `--cpu-backend-for-ml=bnns`, it still requires to load libmkldnn.dylib. It should not do that.
3. When running with switch `--cpu-backend-for-ml=mkldnn`, switching to `SUSTAINED_SPEED` makes chromium crash.
4. When running with `LD_LIBRARY_PATH=`, it requires me to input password every time. @BruceDai mentioned he met same issue.

File: services/ml/public/cpp/ml_switches.cc [hunk 9]
```diff
@@ -2,12 +2,13 @@
...
 // found in the LICENSE file.
 
-#include "services/ml/ml_switches.h"
+#include "services/ml/public/cpp/ml_switches.h"
 
 namespace ml {
 
 namespace switches {
 const char kUseInferenceEngine[] = "use-inference-engine";
+const char kCPUType[] = "cpu-backend-for-ml";
```
Comment (huningxin) [inline services/ml/public/cpp/ml_switches.cc]: suggest naming as "cpu-backend-for-webnn"

File: chrome/browser/about_flags.cc [hunk 10]
```diff
@@ -865,6 +866,16 @@ const FeatureEntry::Choice kAsyncImageDecodingChoices[] = {
...
 };
 
+#if defined(OS_MACOSX)
+const FeatureEntry::Choice kServiceMLCPUBackendChoices[] = {
+    {flag_descriptions::kServiceMLCPUBackendChoicesDefault, "", ""},
+    {flag_descriptions::kServiceMLCPUBackendChoicesForBNNS,
+     ml::switches::kCPUType, "bnns"},
+    {flag_descriptions::kServiceMLCPUBackendChoicesForMKLDNN,
+     ml::switches::kCPUType, "mkldnn"},
+};
```
Comment (fujunwei) [inline chrome/browser/about_flags.cc]: It's only need  one options such "--use-mkldnn-mac" seeing similar usage https://github.com/otcshare/chromium-src/pull/118/commits/89e3e54de6cd3c8b232ac546564550f7980f6469#diff-0bc08e9dff0d633c1458556a0d098a59R264

**Comment by lisa0314** [comment]
@huningxin I have updated the PR. Please help to review.
1. It will not need to require libmkldnn.0.dylib when running bnns.
2. It still need to add `LD_LIBRARY_PATH=` for libiomp.dylib.

**Comment by huningxin** [comment]
@lisa0314 , thanks for the update. 

> * It will not need to require libmkldnn.0.dylib when running bnns.

Nice.

> * It still need to add `LD_LIBRARY_PATH=` for libiomp.dylib.

Please document somewhere, like wiki page, please contact with @BruceDai for instructions.

> * When running with switch --cpu-backend-for-ml=mkldnn, switching to SUSTAINED_SPEED makes chromium crash. I will push new PR to solve that problem since it need to refactor the codes for bnns.

Please file an issue to track it.

**Comment by BruceDai** [comment]
@lisa0314 As we discussed switches on https://github.com/intel/webml-polyfill/issues/744, please update use --use-mkldnn instead of  --cpu-backend-for-ml=mkldnn , thanks.

**Comment by huningxin** [comment]
> --use-mkldnn

I think `--use-mkldnn` is more consistent. The default behavior should be using `bnns`. @BruceDai @lisa0314 , WDYT?

**Comment by lisa0314** [comment]
@BruceDai @huningxin I have update the command line to --use-mkldnn in this PR. Thanks!

File: services/ml/ml_service.cc [hunk 11]
```diff
@@ -22,12 +25,21 @@ MLService::MLService(service_manager::mojom::ServiceRequest request)
 MLService::~MLService() = default;
 
 void MLService::OnStart() {
-#if defined(OS_LINUX) || defined(OS_WIN)
+#if defined(OS_MACOSX)
+  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
+  if (command_line->HasSwitch(switches::kUseMkldnnForMac)) {
+    registry_.AddInterface(base::Bind(&NeuralNetworkImpl::Create));
+  } else if (command_line->HasSwitch(switches::kUseBNNSForMac)) {
```
[the remark this answers is not in the document]
Comment (lisa0314) [reply services/ml/ml_service.cc]: Done.

File: services/ml/ml_switches.cc [hunk 12]
```diff
@@ -12,6 +12,9 @@ const char kUseInferenceEngine[] = "use-inference-engine";
 
 const char kUseDirectML[] = "use-dml";
 
+const char kUseMkldnnForMac[] = "use-mkldnn";
+
+const char kUseBNNSForMac[] = "use-bnns";
```
[the remark this answers is not in the document]
Comment (lisa0314) [reply services/ml/ml_switches.cc]: Done.

File: services/ml/ml_switches.cc [hunk 13]
```diff
@@ -12,6 +12,9 @@ const char kUseInferenceEngine[] = "use-inference-engine";
 
 const char kUseDirectML[] = "use-dml";
 
+const char kUseMkldnnForMac[] = "use-mkldnn";
```
[the remark this answers is not in the document]
Comment (lisa0314) [reply services/ml/ml_switches.cc]: Done.

**Comment by huningxin** [comment]
Looks good. Thanks @lisa0314 .

**Pull request merged**